### PR TITLE
Add keyring support to inv-tool

### PR DESCRIPTION
**Pull request by whd**
It's desirable to avoid storing LDAP credentials in plain text. This patch adds keyring support to inv-tool, akin to boto's, that allows you to store your password in the system keyring. See:

https://pypi.python.org/pypi/keyring

for more about the keyring bindings.

Instead of storing your credentials like this:

```
[authorization]
ldap_username = USERNAME
ldap_password = PASSWORD
```

Do this:

```
[authorization]
ldap_username = USERNAME
keyring = KEYRING_NAME
```

And to add your password to the keyring:

```
$ python
>>> import keyring
>>> keyring.set_password(KEYRING_NAME, USERNAME, PASSWORD)
```

NOTES:

The patch has some documentation change, but doesn't actually say explicitly how to set up the keyring (as this pull request does). The invtool.conf-dist provided looks to be a "fill in these fields" type example configuration file, while this would be of the "do this instead of that" variant. Anyway, docs may need to be updated to make the new functionality clearer.

If you are using Fedora, the python-keyring RPM provided in the standard distribution may be out of date, requiring you to install a newer version via another mechanism.

### CODE REVIEW

**Comment by uberj** [comment]
I like this idea so much that I think we should make the use of the keyring the default option.  I think you should rethink this patch to incorporate bigger changes.

The [keyring](https://pypi.python.org/pypi/keyring#example) page on pip shows an example where the user is prompted to configure their keyring. We could do something like that with invtool. From a users perspective, I would expect something like this:

```
$ invtool
(Configure Invtool)
ldap username: username@mozilla.com
password: ****************
usage: invtool [-h] [--json | --silent | --debug]
               {A,AAAA,CNAME,MX,PTR,SRV,TXT,search,status} ...
invtool: error: too few arguments
$ invtool
usage: invtool [-h] [--json | --silent | --debug]
               {A,AAAA,CNAME,MX,PTR,SRV,TXT,search,status} ...
invtool: error: too few arguments
```

The first invocation of `invtool` should do the `config.has_option('authorization', 'keyring')` check and if it's false, go through the configuration walk through. After the user enters their usename and password, the keyring should be configured as you suggested:

```
keyring.set_password(KEYRING_NAME, USERNAME, PASSWORD)
```

It is also important that the configuration files be modified to contain a new option `keyring` in the `authorization` section so the `has_option` check returns `True` for all later uses:

```
config.set('authorization', 'ldap_username', username)
config.set('authorization', 'keyring', 'ldap')
config.write(open(config_file, 'w'))
```

**Comment by whd** [comment]
tl;dr I am somewhat weary of relying exclusively on a keyring service's existence and fitness. Your call.

Notes/questions before I proceed:
- Do you want to make `keyring` an explicit requirement of `invtool`? 
  
  More below.
- A plain-text password is still supported if and only it is already present.
- Password verification like in the example you mentioned will not be present. 
  
  This would require a network call, and is probably not worth it.
- "ldap" as a keyring service name may be over-broad. 
  
  On the other hand, it could be convenient in the event that multiple client-side programs need access to the same ldap password. I might suggest namespacing it as invtool-ldap or similar if the former is assumed, and leave it as-is if the latter. A third option is to require the user to explicitly name the service to avoid collisions. A fourth is to detect a collision at run-time and warn the user in this case.
- auth failure
  
  The current failure mode with incorrect ldap credentials is to fail when calling out to the API (a JSON decode error). It's not clear that this is due to bad auth, which may be more important to know now that there are multiple ways and places for storing credentials.

Back to the keyring. If we treat `keyring` as an optional dependency and not necessarily available on all platforms, AND we're changing the first-time behavior to require user input in some cases anyway, I would suggest the following behavior. 

On an invocation where neither `ldap_password` nor `keyring` is set and `keyring` is available:

```
$ invtool
(Configure Invtool)
ldap username: username@mozilla.com    # if ldap_password is unset
password: ****************
store password in keyring? (y/n) y
```

If y, attempt to save the password in the keyring and add the `keyring` field to `authorization` (and write the file). Otherwise, don't save the password, keeping it only for the current session. 

When `keyring` is unavailable:

```
$ invtool
(Configure Invtool)
ldap username: username@mozilla.com    # if ldap_username is unset
password: ****************
no keyring module present, store password in plain text? (y/n) y
```

This may be seen as encouraging storing credentials insecurely, but it keeps parity with current functionality. You could even cascade like so:

```
$ invtool
ldap username: username@mozilla.com    # if ldap_username is unset
password: ****************
store password in keyring? (y/n) n
store password in plain text? (y/n) y
```

What say you? I don't mind going with your suggested approach, and making `keyring` required, but in that case I would add a diagnostic message to let the user know that the system keyring is being used to store the password.

**Comment by uberj** [comment]
> Do you want to make keyring an explicit requirement of invtool

Refraining from making it a hard requirement is probably a wise choice; the case of using this for automation on a headless server is a good enough reason.

The (sane) logic I think you arrived at is:
1. Try to use the keychain
2. Try to use the plain text
3. Try to configure the keychain
4. Finally, fail and tell the user to configure the keychain or set a plain text password

As for the keyring service name, I like the idea of initially setting the service name in `config.py-dist` to `invtool-ldap` and mentioning in the docs somewhere that this can be configured.

> Password verification like in the example you mentioned will not be present. The current failure mode with incorrect ldap credentials is to fail when calling out to the API (a JSON decode error). It's not clear that this is due to bad auth, which may be more important to know now that there are multiple ways and places for storing credentials.

The way it should work: If a request is made with invalid credentials an HTTP response with status code 401 will be returned and `invtool` will alert the user accordingly.

I have patches that will fix this, but they are are bundled with other commits that aren't ready for master yet. So for now just assume the user gets their password correct on the first try. I'll handle the 401 eventually.

> `no keyring module present, store password in plain text? (y/n) y`

Indeed, this is scary, let's not help users foot-gun themselves.

It might also be useful to add some output to [status_dispatch.py](https://github.com/uberj/inv-tool/blob/master/invtool/status_dispatch.py) to include whether or not the keychain or a plain text password is being used for auth.

**Comment by whd** [comment]
Here's the updated patch.

I notice you don't use print functions/statements, and instead use a dispatch mechanism to control whether and how to output responses. This facilitates some nice things like --silent and probably --json. However, I am not sure how to reconcile this approach with the notion of requiring user input in some cases during configuration parsing, which happens before actual dispatch begins. This patch adds calls to print, raw_input and getpass.getpass that will not be silenced by --silent.

I somehow got it in my head to introduce as few local variables as possible. There are many calls to config.has_option and config.get_option that could be refactored to use local variables instead.
